test/functional/legacy/094_visual_mode_operators_spec.lua

```diff
@@ -0,0 +1,165 @@
+-- Test for Visual mode and operators.
+--
+-- Tests for the two kinds of operations: Those executed with Visual mode
+-- followed by an operator and those executed via Operator-pending mode. Also
+-- part of the test are mappings, counts, and repetition with the . command.
+
+local helpers = require('test.functional.helpers')
+local feed, insert, source = helpers.feed, helpers.insert, helpers.source
+local clear, execute, expect = helpers.clear, helpers.execute, helpers.expect
+
+-- Vim script user functions needed for some of the test cases.
+local function source_user_functions()
+  source([[
+    function MoveToCap()
+      call search('\u', 'W')
+    endfunction
+    function SelectInCaps()
+      let [line1, col1] = searchpos('\u', 'bcnW')
+      let [line2, col2] = searchpos('.\u', 'nW')
+      call setpos("'<", [0, line1, col1, 0])
+      call setpos("'>", [0, line2, col2, 0])
+      normal! gv
+    endfunction
+  ]])
+end
+
+describe('Visual mode operation', function()
+  setup(clear)
+
+  it('is working', function()
+    insert([[
+      apple banana cherry
+      
+      line 1 line 1
+      line 2 line 2
+      line 3 line 3
+      line 4 line 4
+      line 5 line 5
+      line 6 line 6
+      
+      xxxxxxxxxxxxx
+      xxxxxxxxxxxxx
+      xxxxxxxxxxxxx
+      xxxxxxxxxxxxx]])
+
+    -- Exercise characterwise Visual mode plus operator, with count and repeat.
+    execute('/^apple')
+    feed('lvld.l3vd.')
+
+    -- Same in linewise Visual mode.
+    execute('/^line 1')
+    feed('Vcnewline<esc>j.j2Vd.')
+
+    -- Same in blockwise Visual mode.
+    execute('/^xxxx')
+    feed('<c-v>jlc  <esc>l.l2<c-v>c----<esc>l.')
+
+    -- Assert buffer contents.
+    expect([[
+      a y
+      
+      newline
+      newline
+      
+          --------x
+          --------x
+      xxxx--------x
+      xxxx--------x]])
+  end)
+end)
+
+describe('Visual mode mapping', function()
+  setup(function()
+    clear()
+    source_user_functions()
+  end)
+
+  it('is working', function()
+    insert([[
+      KiwiRaspberryDateWatermelonPeach
+      JambuRambutanBananaTangerineMango]])
+
+    -- Set up Visual mode mappings.
+    execute('vnoremap W /\\u/s-1<CR>')
+    execute('vnoremap iW :<C-U>call SelectInCaps()<CR>')
+
+    -- Do a simple change using the simple vmap, also with count and repeat.
+    execute('/^Kiwi')
+    feed('vWcNo<esc>l.fD2vd.')
+
+    -- Same, using the vmap that maps to an Ex command.
+    execute('/^Jambu')
+    feed('llviWc-<esc>l.l2vdl.')
+
+    -- Assert buffer contents.
+    expect([[
+      NoNoberryach
+      --ago]])
+  end)
+end)
+
+describe('Operator-pending mode mapping', function()
+  setup(function()
+    clear()
+    source_user_functions()
+  end)
```

Comment:
this could be a `before_each` in a single describe block, and it would run before each `it()`. Whereas `setup()` only runs once per describe block.